packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,548 @@
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, 'react-aria-components');
+  let isExpanded = rowProps['aria-expanded'] === true;
+  let hasChildRows = [...state.collection.getChildren!(item.key)]?.length > 1;
+  let level = rowProps['aria-level'];
+
+  let {hoverProps, isHovered} = useHover({
+    isDisabled: !states.allowsSelection && !states.hasAction
+  });
+
+  let {isFocusVisible, focusProps} = useFocusRing();
+  let {checkboxProps} = useGridListSelectionCheckbox(
+    {key: item.key},
+    state
+  );
+
+  let props: TreeItemProps<unknown> = item.props;
+  let renderPropValues = React.useMemo(() => ({
+    ...states,
+    isHovered,
+    isFocusVisible,
+    isExpanded,
+    hasChildRows,
+    level,
+    selectionMode: state.selectionManager.selectionMode,
+    selectionBehavior: state.selectionManager.selectionBehavior
+  }), [states, isHovered, isFocusVisible, state.selectionManager, isExpanded, hasChildRows, level]);
+
+  let renderProps = useRenderProps({
+    ...props,
+    id: undefined,
+    children: item.rendered,
+    defaultClassName: 'react-aria-TreeItem',
+    values: renderPropValues
+  });
+
+  useEffect(() => {
+    if (!item.textValue) {
+      console.warn('A `textValue` prop is required for <TreeItem> elements in order to support accessibility features such as type to select.');
+    }
+  }, [item.textValue]);
+
+
+  let expandButtonProps = {
+    onPress: () => {
+      if (!states.isDisabled) {
+        state.toggleKey(item.key);
+      }
+    },
+    'aria-label': isExpanded ? stringFormatter.format('collapse') : stringFormatter.format('expand'),
+    // TODO: the below actually isn't enough to have keyboard navigation skip over it, we need it to be a span type button but
+    // RAC buttons are always a "button" type element.
+    tabIndex: isAndroid() ? -1 : null
+  };
+
+  let expandButtonRef = useRef<HTMLButtonElement>(null);
+  useEffect(() => {
+    if (hasChildRows && !expandButtonRef.current) {
+      console.warn('Expandable tree rows must contain a expand button so screen reader users can expand/collapse the row.');
+    }
+  // eslint-disable-next-line
+  }, []);
+
+  let children = useCachedChildren({
+    items: state.collection.getChildren!(item.key),
+    children: item => {
+      switch (item.type) {
+        case 'content': {
+          return <TreeRowContent values={renderPropValues} item={item} />;
+        }
+        // Skip item since we don't render the nested rows as children of the parent row, the flattened collection
+        // will render them each as siblings instead
+        case 'item':
+          return <></>;
+        default:
+          throw new Error('Unsupported element type in TreeRow: ' + item.type);
+      }
+    },
+    // TODO: double check if this is the best way to go about making sure TreeRowContent's render props is always up to date
+    dependencies: [renderPropValues]
```

Comment:
open question

Reply:
what are you worried about here? it's a memo'ed object based on a whole list of things that change. so it'll update whenever that list updates.
is it super clear? not particularly, but it's also not hard to figure out

Reply:
I was mainly unsure if this was the proper way to do it or if I was somehow misconfiguring the rendering TreeRowContent. I think the above is correct though.

Reply:
could you pass the renderPropValues through context instead? that way you don't need to worry about the dependencies.

Reply:
oh true, lemme give that a shot